### DISCUSSION TITLE
Idea: Generic type pattern

### DISCUSSION THREAD

**Discussion by JesperTreetop** [dotnet/csharplang#5099] [Language Ideas]
I was looking at an informal static method pattern, let's say a general TryParse form, that could now be formalized as an interface.

```csharp
interface ITryParseable<T> where T : ITryParseable<T> {
	static abstract T? TryParse(string input);
}
```

Having this is great, but like async and Tasks, it all hinges on jumping into a concept at the start. You need some part that's generic to get going, because the only thing that polymorphically can be of that type in C# is a generic type parameter, and the only way to get that is to have a generic method, and now you're back where you started; one way or another you need some sort of trampoline construct, be it Reflection, Expression trees, dynamic, whatever. But what if there was a bridge?

```csharp
Type type; // you already have these from somewhere
string input;
if (type is T where T : ITryParseable<T>) {
	var z = T.TryParse(input);
	// ...
}
```

This doesn't introduce a new variable or variable-like binding. It introduces, for the purposes of the `if`'s scope, the `T` type parameter. `z` is a variable of the type `T` and `T` refers to a type about which is known the constraints in the `where` - just like in a bog standard generic method.

Theoretically, this shouldn't work. When you get to inside the if clause, `T` can be wildly different from one call to the next, and that's not how methods usually work. But when you're trying to get to where you have a generic `T`, it is how you have to make it work by reaching for reflection to call a generic method, say. And it's not dramatically more magic than `yield` or `await` or for that matter capturing variables for closures, all of which the compiler arranges for you by rewriting your code into something rather unrecognizable (for good and ill).

But what would it even desugar to? Let's say this:

```csharp
Type type;
string input;

static void <unmentionable>_Trampoline<T>(string input) where T : ITryParseable<T> {
	var z = T.TryParse(input);
	// ...
}

if (type.GetInterface("Fully.Qualified.ITryParseable`1", ignoreCase:false) is Type iface) {
	var <unmentionable>_trampolineMethodInfo = // .. synthesized by compiler
	var genericMethod = <unmentionable>_trampolineMethodInfo.MakeGenericMethod(new[] { type });
	genericMethod.Invoke(new object[]{ input });
}
```

The above code is possible today and doesn't introduce new behaviors or require new features in the runtime. It's just a synthesis of one way you would solve the bridging.

This has the potential to be slow, but considering the various hoops you have to go through to bridge this gap today, it also has the potential to be made *as fast as it could be* by letting the compiler handle it. It could use whatever fast paths through reflection or alternative mechanisms are available, and could be made to survive linking because it's not open-season reflection, it comes from a construct whose behavior is narrow, defined and known.

It does involve hoisting code out of that `if`'s innards and placing it somewhere else. But, again, that's the bread and butter of `async`, iterators, anonymous methods or lambdas already, and if it can happen there to make those scenarios easier, why not here to make this scenario easier?

**Comment by svick** [comment]
> But, again, that's the bread and butter of async, iterators, anonymous methods or lambdas already, and if it can happen there to make those scenarios easier, why not here to make this scenario easier?

One potential answer is that it's because the existing features significantly simplify very common coding patterns. This proposal would modestly simplify a pattern that's probably going to be relatively uncommon.

**Reply by JesperTreetop** [reply]
Relatively uncommon compared to async methods, sure. But a lot of (invasive) effort just went into making `static abstract` in interfaces happen, and since there's nothing other than a generic type parameter you can use to call those methods, getting into the model is a problem compared to calling instance-level members on interfaces. In other words, proportionally to that feature, it wouldn't be that uncommon. 

This is a cliff that basically everyone who wants to use the generality of the static interface methods will run into. You can't tell me that if you had to do the equivalent reflection incantations to call an interface *instance* method or property, that it wouldn't affect how you used them.

**Reply by HaloFour** [reply]
Given that static abstract methods are only in preview and they represent a building blocks for "shapes" I'd rather see where the design of those features is headed before language features based on reflection are suggested.

**Reply by JesperTreetop** [reply]
I'm looking forward to shapes. And for what it's worth, I think a language feature that says what you want to accomplish and lets the compiler arrange for it to happen is much less likely to mire codebases with reflection code than *having everyone write reflection code*.
